### PR TITLE
Adding benchmark for HttpClient instrumentation

### PR DESCRIPTION
**Pull request by eddynaka**
|           Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| SimpleHttpClient | 168.7 us | 3.26 us | 3.35 us | 1.4648 |     - |     - |   1.84 KB |
|                   InstrumentedHttpClient | 191.9 us | 3.81 us | 5.81 us | 2.1973 |     - |     - |   4.68 KB |
| InstrumentedHttpClientWithParentActivity | 189.1 us | 2.69 us | 2.51 us | 2.4414 |     - |     - |   5.97 KB |

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: test/Benchmarks/Instrumentation/HttpInstrumentationBenchmark.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        public void GlobalCleanup()
+        {
+            this.httpClient.Dispose();
+            this.source.Dispose();
+        }
+
+        [Benchmark(Baseline = true)]
+        public async Task SimpleHttpClient()
+        {
+            var httpResponse = await this.httpClient.GetAsync("https://opentelemetry.io/");
```
Comment (cijothomas) [inline test/Benchmarks/Instrumentation/HttpInstrumentationBenchmark.cs:50]: can we setup own in-proc server? 
or how about using a special handler in the httpclient, which simply returns 200 OK. 
(We need to make sure the diagnostichandler is still part of httpclient's handler pipeline, as thats the one which emits DiagnosticSource events)

File: test/Benchmarks/Instrumentation/HttpInstrumentationBenchmark.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            httpResponse.EnsureSuccessStatusCode();
+        }
+
+        [Benchmark]
+        public async Task InstrumentedHttpClient()
+        {
+            using var openTelemetry = Sdk.CreateTracerProviderBuilder()
```
Comment (cijothomas) [inline test/Benchmarks/Instrumentation/HttpInstrumentationBenchmark.cs:57]: does this get executed in every benchmark call? I think this should be a one time setup..

Comment (eddynaka) [reply test/Benchmarks/Instrumentation/HttpInstrumentationBenchmark.cs:57]: i was thinking about that. If I move to the constructor, for example, the SimpleHttpClient will be monitored as well. It's why I added to the method itself.

Comment (eddynaka) [reply test/Benchmarks/Instrumentation/HttpInstrumentationBenchmark.cs]: changed to a separate file and moved to global setup method.

File: test/Benchmarks/Instrumentation/HttpInstrumentationBenchmark.cs (hunk 1 above)
Comment (eddynaka) [reply test/Benchmarks/Instrumentation/HttpInstrumentationBenchmark.cs]: used the in-proc server approach. thanks for the guidance.

File: test/Benchmarks/Instrumentation/HttpInstrumentationBenchmark.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        public async Task InstrumentedHttpClient()
+        {
+            var httpResponse = await this.httpClient.GetAsync(this.url);
+            httpResponse.EnsureSuccessStatusCode();
+        }
+
+        [Benchmark]
+        public async Task InstrumentedHttpClientWithParentActivity()
+        {
+            using var parent = this.source.StartActivity("incoming request", ActivityKind.Server);
```
Comment (cijothomas) [inline test/Benchmarks/Instrumentation/HttpInstrumentationBenchmark.cs:82]: can you move "incoming request" to const string.

Comment (eddynaka) [reply test/Benchmarks/Instrumentation/HttpInstrumentationBenchmark.cs]: done.

**Comment by cijothomas** [comment]
> Method	Mean	Error	StdDev	Gen 0	Gen 1	Gen 2	Allocated
> SimpleHttpClient	168.7 us	3.26 us	3.35 us	1.4648	-	-	1.84 KB
> InstrumentedHttpClient	191.9 us	3.81 us	5.81 us	2.1973	-	-	4.68 KB
> InstrumentedHttpClientWithParentActivity	189.1 us	2.69 us	2.51 us	2.4414	-	-	5.97 KB
>

The numbers show we have work to do in optimizing the instrumentations!

File: test/Benchmarks/Benchmarks.csproj [hunk 4]
```diff
@@ -18,6 +18,7 @@
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Jaeger\OpenTelemetry.Exporter.Jaeger.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Zipkin\OpenTelemetry.Exporter.Zipkin.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry\OpenTelemetry.csproj" />
+    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Instrumentation.Http\OpenTelemetry.Instrumentation.Http.csproj" />
```
Comment (reyang) [inline test/Benchmarks/Benchmarks.csproj:21]: nit: sort alphabetically

File: test/Benchmarks/Instrumentation/UnInstrumentedHttpClientBenchmark.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
+﻿// <copyright file="UnInstrumentedHttpClientBenchmark.cs" company="OpenTelemetry Authors">
```
Comment (reyang) [inline test/Benchmarks/Instrumentation/UnInstrumentedHttpClientBenchmark.cs:1]: UnInstrumented -> Uninstrumented (lowercase `i`).

File: test/Benchmarks/Instrumentation/UnInstrumentedHttpClientBenchmark.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+using System;
+using System.Net.Http;
+using System.Threading.Tasks;
+using BenchmarkDotNet.Attributes;
+using OpenTelemetry.Tests;
+
+namespace Benchmarks.Instrumentation
+{
+    [MemoryDiagnoser]
+    public class UnInstrumentedHttpClientBenchmark
```
Comment (reyang) [inline test/Benchmarks/Instrumentation/UnInstrumentedHttpClientBenchmark.cs:26]: I'm not good at naming things. Curious to know if folks think we should call it `UninstrumentedSomething`.
It reminded me of an old joke where a C# developer asked a question in a C++ email group about "unmanaged code", immediately lots of C++ developers replied "don't call it unmanaged, use the word `native`".

File: test/Benchmarks/Tracing/OpenTelemetrySdkBenchmarksActivity.cs [hunk 7]
```diff
@@ -25,30 +25,39 @@ public class OpenTelemetrySdkBenchmarksActivity
...
         private readonly ActivityContext parentCtx = new ActivityContext(ActivityTraceId.CreateRandom(), ActivitySpanId.CreateRandom(), ActivityTraceFlags.None);
         private readonly string parentId = $"00-{ActivityTraceId.CreateRandom()}.{ActivitySpanId.CreateRandom()}.00";
+        private TracerProvider openTelemetrySdk;
 
-        public OpenTelemetrySdkBenchmarksActivity()
+        [GlobalSetup]
+        public void GlobalSetup()
         {
-            var openTel = Sdk.CreateTracerProviderBuilder()
+            this.openTelemetrySdk = Sdk.CreateTracerProviderBuilder()
```
Comment (reyang) [inline test/Benchmarks/Tracing/OpenTelemetrySdkBenchmarksActivity.cs:33]: Probably call it ``traceProvider`?

Comment (reyang) [reply test/Benchmarks/Tracing/OpenTelemetrySdkBenchmarksActivity.cs:33]: @cijothomas

Comment (eddynaka) [reply test/Benchmarks/Tracing/OpenTelemetrySdkBenchmarksActivity.cs:33]: just updated! thanks!

File: test/Benchmarks/Instrumentation/UnInstrumentedHttpClientBenchmark.cs (hunk 6 above)
Comment (eddynaka) [reply test/Benchmarks/Instrumentation/UnInstrumentedHttpClientBenchmark.cs]: For now, I changed to `UninstrumentedSomething`.

File: test/Benchmarks/Benchmarks.csproj (hunk 4 above)
Comment (eddynaka) [reply test/Benchmarks/Benchmarks.csproj:21]: done!

**Pull request merged**